### ISSUE TITLE
Ensure Huggingface, Torch Packages Up-to-Date with Necessary Improvements/Fixes

### ISSUE THREAD

**Issue by fabianlim** [foundation-model-stack/fms-hf-tuning#60]
@hickeyma @Ssukriti @tedhtchang

The platform is build upon the `torch` and `huggingface` ecosystem whereby there are constant fixes that are pushed every release. 
- sometimes we also require to be on nightly pre-releases as there is a biweekly-monthly lag for each release.
- The fear of upgrading packages comes with breaking APIs, which will severely impact any CI/CD testing. On the other hand, fixes improve memory consumption and training speed, and will be a good idea to regress the codebase with the latest nightlies.

#### Desiderata
1. Have a ground source of truth for package versions. A good choice would be `pyproject.toml` that can be used to also address #56 , #57 
2. Have a system whereby nightlies (or new releases) can be constantly regressed with new code changes (a good choice would be to have one such regression for each new PR).
   - will be a good idea to also consider some means to disable regressions, for PRs that do not update source code (e.g., documentation updates). 
3. Have a means to _freeze package versions_ during CD/CI testing, so that if a test fails, we know exactly on what versions it has failed for. And vice-versa, if a commit is known to work, we know exactly for what versions it worked for.

#### NPM addresses these desiderata for `node`

See [here](https://support.deploybot.com/article/131-why-developers-should-use-npm-ci-instead-of-npm-install-and-its-benefits):
- `npm install` will occur from `package.json` with semvar versioning that tells what kind of upgrades (e.g., minor, patch) are allowed. 
   * during `npm install` the packages will be resolved to install the _latest_ packages that satisfy the constraints. 
   * the final versions of packages installed will be output in `package-lock.json`.
   * the `package-lock.json` is typically checked to have a record of what exact package versions are working.
 - during CI/CD, all tests are run only after `npm ci`; note that this differs from `npm install`, in that it installs from `package-lock.json`. __Thus there is no package resolution during testing__. This ensures if the tests pass, we know exactly what package versions they pass for.

#### Solutions for Python

There are at least two possibilities that we know of:
1. Poetry has `poetry.lock`, that works together with `pyproject.toml`. The [poetry flow as described here](https://python-poetry.org/docs/basic-usage/#installing-with-poetrylock) is completely analogue with the `npm` flow described above.
2. Using `pip-compile` to generate `requirements.txt` from `pyproject.toml` and checking that in as the lock file.

**Comment by Ssukriti** [comment]
@gkumbhat is thinking about how to do dependency management and will update when done

**Comment by VassilisVassiliadis** [comment]
@tedhtchang we have a couple of suggestions in this issue for ways to version control the versions of python dependencies.

For what it's worth, I started playing poetry on a fork of fms-hf-tuning: https://github.com/VassilisVassiliadis/fms-hf-tuning/blob/2d3b3a040b881a5ee21e2a81bf82134ca5fe6bbb/pyproject.toml#L5-L49

The resulting poetry lock file is here: https://github.com/VassilisVassiliadis/fms-hf-tuning/blob/2d3b3a040b881a5ee21e2a81bf82134ca5fe6bbb/poetry.lock

**Comment by tedhtchang** [comment]
@VassilisVassiliadis The  poetry framework is a good suggestion. Seems like we may need to change how we run test, release, and build packages. Would you like to create a PR ?

**Comment by VassilisVassiliadis** [comment]
I'll give it a go
